tdrs-backend/tdpservice/search_indexes/parsers/preparser.py

```diff
@@ -0,0 +1,262 @@
+    datafile.seek(0)
+    row = datafile.readline()
+    datafile.seek(0)  # reset file pointer to beginning of file
+
+    # datafile when passed via redis/db is a FileField which returns bytes
+    if isinstance(row, bytes):
+        row = row.decode()
+
+    if get_record_type(row) != 'HE':
+        raise ValueError('First line in file is not recognized as a valid header.')
+    elif len(row) != 24:
+        logger.debug("row: '%s' len: %d", row, len(row))
+        return False, {'preparsing': 'Header length incorrect.'}
+    row = row.strip('\n')
+
+    return True, row
+
+def get_trailer_row(datafile):
+    """Alters the trailer row into usable string."""
+    # certify/transform input row to be correct form/type
+
+    # Don't want to read whole file, just last line, only possible with binary
+    # Credit: https://openwritings.net/pg/python/python-read-last-line-file
+    # https://stackoverflow.com/questions/46258499/
+    try:
+        datafile.seek(-2, os.SEEK_END)  # Jump to the second last byte.
+        while datafile.read(1) != b'\n':  # Check if new line.
+            datafile.seek(-2, os.SEEK_CUR)   # Jump two bytes back
+    except OSError:  # Either file is empty or contains one line.
+        datafile.seek(0)
+        return False, {'preparsing': 'File too short or missing trailer.'}
+
+    # Having set the file pointer to the last line, read it in.
+    row = datafile.readline().decode()
+    datafile.seek(0)  # Reset file pointer to beginning of file.
+
+    logger.info("Trailer row: '%s'", row)
+    if get_record_type(row) != 'TR':
+        raise ValueError('Last row is not recognized as a trailer row.')
+    elif len(row) != 24:
+        logger.debug("row: '%s' len: %d", row, len(row))
+        return False, {'preparsing': 'Trailer length incorrect.'}
+    row = row.strip('\n')
+
+    return True, row
+
+def preparse(data_file, data_type, section):
+    """Validate metadata then dispatches file to appropriate parser."""
+    if isinstance(data_file, DataFile):
+        logger.debug("Beginning preparsing on '%s'", data_file.file.name)
+        datafile = data_file.file  # do I need to open() this?
+    elif isinstance(data_file, BufferedReader):
+        datafile = data_file
+    else:
+        logger.error("Unexpected datafile type %s", type(data_file))
+        raise TypeError("Unexpected datafile type.")
+
+    header_preparsed, row = get_header_row(datafile)
+    if header_preparsed is False:
+        return False, row
+    # logger.debug("Header: %s", row)
+
+    header_is_valid, header_validator = validate_header(row, data_type, section)
+    if isinstance(header_validator, Exception):
+        raise header_validator
+
+    trailer_preparsed, row = get_trailer_row(datafile)
+    if trailer_preparsed is False:
+        return False, row
+    trailer_is_valid, trailer_validator = validate_trailer(row)
+    if isinstance(trailer_validator, Exception):
+        raise trailer_validator
```

Comment:
Since it appears these returns are only used by the tests, this instance validation checking could be done in the validate functions. This would clean up this function.

Reply:
These are only used by the tests because it's brand new code -- nothing else yet depends on the (pre)parser. I made [another comment](https://github.com/raft-tech/TANF-app/pull/2312/files#r1055601744) lower down about us deciding on the return structure in 1354 and I think this could/should be a team-wide discussion as we start integrating the parser into the parserLog and potentially other aspects.